### ISSUE TITLE
RFE: Add support for Egress Gateway in Istio

### ISSUE THREAD

**Issue by panoslith** [rancher/rancher#23517]
**What kind of request is this (question/bug/enhancement/feature request):**
Feature request

**Details**
Current Rancher integration with Istio does not support [Istio's Egress Gateway](https://istio.io/docs/tasks/traffic-management/egress/egress-gateway/). It would be useful to be supported, as it would limit which services can or should access external networks and enfornce secure control of egress traffic. 

gzrancher/rancher#6602

**Comment by cloudnautique** [comment]
The current Rancher Istio chart will need to bring in additional CRDs.
Add images to the rancher-images.txt files.

The egress deployment options need to be exposed in the deployment interface in Rancher UI.
egress options need to support tolerations, and node selection via labels. 

Rancher UI should have egress gateway tab in UI to edit the CRDs from the UI.

**Comment by cjellick** [comment]
@rancher-max the options to turn this on are in the chart. Can you figure out what flags are needed to enable and then do a quick smoke test to assert basic functionality of the egress? Post your findings back here. You can ping @janeczku or @cloudnautique with any specific follow up questions.

**Comment by rancher-max** [comment]
### Possible in current state by setting a few Custom Answers upon enabling Istio:
```
gateways.enabled=true
gateways.istio-egressgateway.enabled=true
global.proxy.accessLogFile=/dev/stdout
```
The first two answers there enable Egress Gateway. Due to default values, this also will enable Ingress Gateway. If ingress functionality is not desired, then you must also supply the answer: `gateways.istio-ingressgateway.enabled=false` The other answer supplied there allows us to see our egress traffic using `kubectl logs -l istio=egressgateway -c istio-proxy -n istio-system | tail`

---
### (below is just information to run a simple smoketest. Refer to Istio's official docs: https://istio.io/docs/tasks/traffic-management/egress/egress-gateway/)

Using kubectl is the easiest way, but this is also possible through the UI (see below):

First, ensure you have a running pod that has curl enabled (apply https://raw.githubusercontent.com/istio/istio/release-1.4/samples/sleep/sleep.yaml for a simple workload with curl). Then, use "Import Yaml" to deploy the ServiceEntry. The Gateway, DestinationRule, and VirtualService can all be deployed by manually filling in the correct values on the corresponding Resources > Istio page. After all of these have been created, navigate to the workload with curl, select "Execute Shell", and run `curl -sL -o /dev/null -D - http://edition.cnn.com/politics`. 

After completing these steps, the output of `kubectl logs -l istio=egressgateway -c istio-proxy -n istio-system | tail` should contain something similar to: `[2020-01-28T17:08:20.532Z] "GET /politics HTTP/2" 301 - "-" "-" 0 0 14 1 "10.42.1.145" "curl/7.64.0" "67c0036c-a69b-4a05-863e-9ce20b393406" "edition.cnn.com" "151.101.193.67:80" outbound|80||edition.cnn.com - 10.42.1.146:80 10.42.1.145:34340 - -` which indicates Istio Egress Gateway is working successfully.

**Comment by janeczku** [comment]
@rancher-max ! thanks for verifying and documenting the steps required. Will pass on the information. Has this been verified on the 1.4.300 version of the charts?

**Comment by rancher-max** [comment]
Yep I validated using latest release (Rancher v2.3.5) and Istio version available there (1.4.300)

**Comment by bmdepesa** [comment]
rancher/rancher:master-head c428bc9feb4b11539bacdfa13f173f22a593eaa4

* Deploy cluster
* Deploy v2 Istio w/ Egress Gateway enabled
* Deploy a test workload `https://raw.githubusercontent.com/istio/istio/release-1.4/samples/sleep/sleep.yaml`
* Following the directions from https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/
  * Create an egress gateway, service entry, virtual service, destination rule
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: cnn
spec:
  hosts:
  - edition.cnn.com
  ports:
  - number: 80
    name: http-port
    protocol: HTTP
  - number: 443
    name: https
    protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-egressgateway
spec:
  selector:
    istio: egressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - edition.cnn.com
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: egressgateway-for-cnn
spec:
  host: istio-egressgateway.istio-system.svc.cluster.local
  subsets:
  - name: cnn
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: direct-cnn-through-egress-gateway
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - istio-egressgateway
  - mesh
  http:
  - match:
    - gateways:
      - mesh
      port: 80
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: cnn
        port:
          number: 80
      weight: 100
  - match:
    - gateways:
      - istio-egressgateway
      port: 80
    route:
    - destination:
        host: edition.cnn.com
        port:
          number: 80
      weight: 100
```
  *  In the test pod: `curl -sL -o /dev/null -D - http://edition.cnn.com/politics`
```
...
HTTP/1.1 301 Moved Permanently
...
location: https://edition.cnn.com/politics
...
HTTP/2 200
Content-Type: text/html; charset=utf-8
... 
```
With Kiali & Prometheus configured, we can see the traffic exiting the egress gateway:
![Screen Shot 2020-09-01 at 12 10 13 PM](https://user-images.githubusercontent.com/21316915/91905679-c0db5480-ec5b-11ea-857b-bc26ebd2e3e6.png)

**Comment by davidxianggao** [comment]
@rancher-max can you please address a bit detail how to use those 3 following values to enable an egress gateway plus keeping the existing ingress gateway coming with Rancher? Do you use the command istioctl with them? If you don't mind, can you show the full command line?

gateways.enabled=true
gateways.istio-egressgateway.enabled=true
global.proxy.accessLogFile=/dev/stdout

**Comment by bmdepesa** [comment]
@davidxianggao these values are supplied to the custom answers section in the Istio deployment page in Rancher 2.4.8 and previous - for Istio 1.4.x/1.5.x:

![Screen Shot 2020-10-09 at 7 53 17 AM](https://user-images.githubusercontent.com/21316915/95598221-85733900-0a04-11eb-8cd4-9ff208437b59.png)

For Rancher 2.5.0+, you can enable Egress Gateways by default in the Istio app install page, and use a custom overlay to customize your egress gateways as needed.

Neither of these options directly use istioctl.

**Comment by davidxianggao** [comment]
@bmdepesa Thank you for your sharing screenshot. I've added those 3 custom answers. The egress gateway got created as expected. Much appreciate you and @rancher-max. This egress thing cost us a few weeks until we see this post.

Also, I notice that the custom answer of "gateways.enabled=true" disappeared after I enabled them, even though both of egree and ingress gateway and the rest two custom answers still exists now. It's not a big deal just wanna confirm if it is expected too. My Rancher version is 2.4.6 and Istio is 1.5.9
